### DISCUSSION TITLE
Difference between mqtt-export and external-mqtt-trigger

### DISCUSSION THREAD

**Discussion by neerajasjawali** [edgexfoundry/community#60] [Q&A] [answered]
Hi all,

I am working on reading the uart sensors data using edgex (jakarta). I am testing the edgex message-bus by using all the types of buses supported by edgex. Firstly, I tested with default message-bus used by edgex, i,e. redis pub/sub. So, tested reading the sensors data and publishing and subscribing the data is possible.But its not efficient to use redis pub/sub for publishing the larger data. Next, I modified the app-service (mqtt-export) to use mqtt for pub/sub, the app-service and device-service are able to connecting to broker. At sometimes the services get disconnected and reconnect with the broker, due to which there is data loss happening. Lastly, I tried working with external-mqtt-trigger for pub/sub. Based on my testing I could say that external-mqtt-trigger is used to get (subscribe) the   commands/data from the publisher (cloud). 

I would appreciate the help resolving the issues 
1) The difference between edgex app-service "mqtt-export" and "external-mqtt-trigger"? 
2) How can I make the services to never get disconnected from the broker by its own?

Thanks in advance

**Comment by lenny-goodell** [comment]
Please read the documentation on triggers and export here to understand the difference.
https://docs.edgexfoundry.org/2.1/microservices/application/Triggers/
https://docs.edgexfoundry.org/2.1/microservices/application/BuiltIn/#export

**Comment by lenny-goodell** [comment, marked as answer]
The KeepAlive setting for the MQTT connections can help with disconnecting. This can be set in configuration for both MQTT Export and MQTT Trigger. Please refer to MQTT external specs for appropriate values.

**Reply by neerajasjawali** [reply]
Thank you @lenny-intel
